gravitee-apim-gateway/gravitee-apim-gateway-tcp/src/main/java/io/gravitee/gateway/tcp/reactive/TcpResponse.java

```diff
@@ -0,0 +1,85 @@
+import io.gravitee.gateway.api.http.HttpHeaders;
+import io.gravitee.gateway.reactive.api.context.GenericExecutionContext;
+import io.gravitee.gateway.reactive.api.context.InternalContextAttributes;
+import io.gravitee.gateway.reactive.core.context.AbstractResponse;
+import io.reactivex.rxjava3.core.Completable;
+import io.reactivex.rxjava3.core.Flowable;
+import io.vertx.rxjava3.FlowableHelper;
+import io.vertx.rxjava3.core.streams.WriteStream;
+
+/**
+ * This class represents the outbound TCP connection. It serves as an end of the tcp-server -> backend -> tcp-server round trip and handles the bidirectional data flow creation.
+ *
+ * @author Benoit BORDIGONI (benoit.bordigoni at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class TcpResponse extends AbstractResponse {
+
+    private final TcpRequest request;
+
+    public TcpResponse(TcpRequest request) {
+        super();
+        this.request = request;
+        this.headers = HttpHeaders.create();
+        this.trailers = HttpHeaders.create();
+    }
+
+    /**
+     * This method send upstream traffic from the client via {@link TcpRequest#chunks()} towards the backend (prepared by the endpoint)
+     * and simultaneously reads downstream traffic from the backend via {@link #chunks()} back to the client using a vertx {@link io.vertx.rxjava3.core.streams.Pipe}
+     * to handle automatically the backpressure. Using of chunks method allow policies to manipulate buffers in policy during request and response phases.
+     * @param ctx the execution context
+     * @return a Completable that performs the piping at subscription time.
+     */
+    @Override
+    public Completable end(GenericExecutionContext ctx) {
+        return Completable
+            .defer(() -> {
+                WriteStream<io.vertx.rxjava3.core.buffer.Buffer> backendWriteStream = ctx.getInternalAttribute(
+                    InternalContextAttributes.ATTR_INTERNAL_TCP_BACKEND_RESPONSE_STREAM_INTERNAL_ATTRIBUTE
+                );
+                return Completable.mergeArray(
+                    // Read request chunks and write to backendSocket
+                    createReadStream(request.chunks()).rxPipeTo(backendWriteStream),
+                    // Read response chunks and write to proxy socket
+                    createReadStream(this.chunks()).rxPipeTo(request.getWriteStream())
+                );
+            })
+            .doFinally(() -> this.ended = true);
+    }
+
+    private io.vertx.rxjava3.core.streams.ReadStream<io.vertx.rxjava3.core.buffer.Buffer> createReadStream(final Flowable<Buffer> chunks) {
+        return io.vertx.rxjava3.core.streams.ReadStream.newInstance(
+            FlowableHelper.toReadStream(
+                chunks.map(buffer -> {
+                    byte[] bytes = new byte[buffer.getNativeBuffer().readableBytes()];
+                    buffer.getNativeBuffer().readBytes(bytes);
+                    return io.vertx.core.buffer.Buffer.buffer(bytes);
```

Comment:
`chunks.map(buffer -> io.vertx.core.buffer.Buffer.buffer(buffer.getBytes()))` ?

Reply:
That was my first implementation, but nope, it is deprecated for removal in vertx 5

Reply:
🤔 in both cases it uses `io.vertx.core.buffer.Buffer.buffer(byte[])` 
What I'm proposing is to replace 

```java
byte[] bytes = new byte[buffer.getNativeBuffer().readableBytes()];
buffer.getNativeBuffer().readBytes(bytes);
```

with

```java
buffer.getBytes()
```